### PR TITLE
Upgrade to webpack@5

### PR DESCRIPTION
**Pull request by ixrock**
This also fixing nasty caching bug with UI files after combination(s) of `make clean && make dev` / page reload / `git checkout` branches / etc.

**TODO:**

- [x] fix unit tests
- [x] fix integration tests
- [x] fix warnings in compilation of `main` process _(types must be imported separately in files participating in extension-api re-exports)_ 
- [x] fix incorrect import warnings in compilation of `renderer` process
- [x] use `@pmmmwh/react-refresh-webpack-plugin` with `react-refresh-typescript` to support HMR in most cases

### CODE REVIEW

**Comment by Nokel81** [comment]
Shouldn't think be marked as a chore?

**Comment by ixrock** [comment]
> Shouldn't think be marked as a chore?

whatever :)

File: webpack.renderer.ts
Comment (ixrock) [inline webpack.renderer.ts:49]: Not supported in `webpack@5`. What is was used for? @jakolehm maybe you know?

File: src/common/catalog-entities/kubernetes-cluster.ts
Comment (Nokel81) [inline src/common/catalog-entities/kubernetes-cluster.ts:138]: Woot! 🎉

File: src/main/index.ts
Comment (Nokel81) [inline src/main/index.ts]: Can we not have these all on dual lines?

File: src/main/window-manager.ts
Comment (Nokel81) [inline src/main/window-manager.ts]: I don't think this needs to be `@observable`

File: src/renderer/components/dock/terminal/terminal.ts
Comment (Nokel81) [inline src/renderer/components/dock/terminal/terminal.ts]: Is it possible to move this to the `HTML` file?

File: webpack.main.ts
Comment (Nokel81) [inline webpack.main.ts:24]: Why was this switched?

File: src/main/index.ts
Comment (ixrock) [reply src/main/index.ts]: Changes reverted. It was probably auto-formatting files via [.idea](https://github.com/lensapp/lens/tree/master/.idea) settings. Why do we even need to commit them to the repo? @jansav any clue?

File: webpack.main.ts
Comment (ixrock) [reply webpack.main.ts:24]: Because it is the only place where `isProduction` was used when `isDevelopment` used in multiple places.
Also it's done to streamline code with other `webpack` configurations.

File: src/renderer/components/dock/terminal/terminal.ts
Comment (ixrock) [reply src/renderer/components/dock/terminal/terminal.ts:31]: If we have settings to change terminal's font - than not. (if it's not always bundled with the app)
But I have to double check this..

File: webpack.main.ts
Comment (Nokel81) [reply webpack.main.ts:24]: Sounds good, I didn't notice that you also removed `isProduction` as a variable but reducing that set is good.

File: src/renderer/components/dock/terminal/terminal.ts
Comment (ixrock) [reply src/renderer/components/dock/terminal/terminal.ts]: Preloading fonts moved to `template.html` and `fonts.scss`

File: src/main/window-manager.ts
Comment (ixrock) [reply src/main/window-manager.ts]: Removed.

File: src/main/menu/menu.ts
Comment (ixrock) [inline src/main/menu/menu.ts]: Not used anywhere.

File: src/common/catalog-entities/kubernetes-cluster.ts
Comment (ixrock) [reply src/common/catalog-entities/kubernetes-cluster.ts:138]: By default imported `*.svg` files contain data-url (string) that can be consumed normally by `` or with `<Icon svg/>` as inline/raw xml icon.

File: src/main/menu/menu.ts
Comment (Nokel81) [reply src/main/menu/menu.ts]: It should be, for the registration of menu items since we don't support adding new menu items.

Comment (ixrock) [reply src/main/menu/menu.ts]: Well, it was used before. Dunno who was refactored menu that it now has some leftovers. 
Anyway, this is out of scope.

Comment (Nokel81) [reply src/main/menu/menu.ts]: I think it is out of scope to remove it.

Comment (ixrock) [reply src/main/menu/menu.ts]: Nope. This is just like removing a comment since it's not used / doesn't touch any other code.

Comment (Nokel81) [reply src/main/menu/menu.ts]: Disagree, an upgrade PR like this should be only doing the minimal work to get it working, this should be a separate PR.

Comment (ixrock) [reply src/main/menu/menu.ts]: Okay, I'll revert it, but just want to say.. With that frenetic attitude in PR's you probably gonna work alone on this project ~(maybe this is your goal)~.. Just see how other projects has quick PR reviews without PITA..

**Comment by ixrock** [comment]
@lensapp/lens-maintainers PTAL, ready for testing & reviews.

_HMR demo:_

https://user-images.githubusercontent.com/6377066/153879443-bea48fa6-70aa-490f-bb65-77d3834a32e2.mov

File: extensions/kube-object-event-status/package.json
Comment (Nokel81) [inline extensions/kube-object-event-status/package.json]: Can you revert this change?

File: extensions/kube-object-event-status/package.json
Comment (Nokel81) [inline extensions/kube-object-event-status/package.json:21]: We shouldn't use `latest` here because it will break if there is a mismatch of versions between this file and `../../package.json`

File: src/renderer/components/+network-ingresses/ingress-details.tsx
Comment (Nokel81) [inline src/renderer/components/+network-ingresses/ingress-details.tsx:18]: Wondering out loud, is this down automatically? (Because it would be great if it was)

File: webpack.extensions.ts
Comment (Nokel81) [inline webpack.extensions.ts]: Shouldn't we switch to just using `isDevelopment` here too?

File: src/renderer/components/+network-ingresses/ingress-details.tsx
Comment (ixrock) [reply src/renderer/components/+network-ingresses/ingress-details.tsx:18]: No idea.. but those warnings just in compilation process and probably could be ignored.

File: webpack.extensions.ts
Comment (ixrock) [reply webpack.extensions.ts]: Yes, why not.

File: extensions/kube-object-event-status/package.json
Comment (ixrock) [reply extensions/kube-object-event-status/package.json:21]: Why it will break when it's not used in "production" / inside packed extension?
I think only `ts-loader` must be compatible with local `webpack` package.

File: extensions/kube-object-event-status/package.json
Comment (ixrock) [reply extensions/kube-object-event-status/package.json]: Reverted.

File: extensions/kube-object-event-status/package.json
Comment (Nokel81) [reply extensions/kube-object-event-status/package.json:21]: We do run `npm install` on extensions, even bundled extensions.

Comment (ixrock) [reply extensions/kube-object-event-status/package.json:21]: When we do? So maybe we should move it to `peerDependencies` then?
Anyway I think it should be handled from parent's package version, otherwise it's just copy-pasting all around. Can we use actual versions from app's `package.json` on extensions install?

Comment (Nokel81) [reply extensions/kube-object-event-status/package.json:21]: https://github.com/lensapp/lens/blob/master/src/extensions/extension-discovery/extension-discovery.ts#L383-L406

This is where we do the install calls

Comment (ixrock) [reply extensions/kube-object-event-status/package.json:21]: Well, it's still unclear how this installing dependency handle unpacking extension:

```ts
await this.dependencies.installExtension(extension.absolutePath);
```

In general it feels wrong if compiled extension require matching own development dependencies with the  app (in case of `react`, `mobx` and some other packages we're sharing them via global env because version matching there is important for app's work. @jakolehm @nevalla @panuhorsmalahti wdyt?

Comment (jakolehm) [reply extensions/kube-object-event-status/package.json:21]: @Nokel81 @ixrock Lens installs extension npm package without development dependencies -> extension dev dependencies don't need to match same versions on Lens side.

Comment (jakolehm) [reply extensions/kube-object-event-status/package.json:21]: Note: using `latest` version might still be a bit risky (in general).

**Review by jakolehm — APPROVED** [review]
LGTM

File: extensions/kube-object-event-status/package.json
Comment (ixrock) [reply extensions/kube-object-event-status/package.json:21]: I think it's not more risky than just having outdated specific version == in both cases I assume it's easy to spot/fix/use what is needed per extension package.

File: src/common/catalog-entities/kubernetes-cluster.ts
Comment (aleksfront) [reply src/common/catalog-entities/kubernetes-cluster.ts:138]: Amazing!

**Comment by aleksfront** [comment]
Found 2 terminal-related issues:
1. Fonts don't preload properly in terminal window.
broken fonts

2. An error occured on every terminal tab close:
```
Uncaught TypeError: Cannot read properties of null (reading 'url')
```

https://user-images.githubusercontent.com/9607060/154052621-d1aeeeb4-ccc5-4793-98f5-d13ec72b8076.mov

**Comment by Nokel81** [comment]
@aleksfront that second issue is a known issue that exists on master. I don't think it should block this PR

**Review by Nokel81 — APPROVED** [review]
Except for the font issue. This looks good to me.

**Comment by ixrock** [comment]
> Except for the font issue. This looks good to me.

fix is already done (gonna create new PR very soon)

**Pull request merged**
